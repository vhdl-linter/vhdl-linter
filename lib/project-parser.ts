import { FSWatcher, watch } from 'chokidar';
import { EventEmitter } from 'events';
import { promises } from 'fs';
import { join, sep } from 'path';
import { cwd } from 'process';
import { OContext, OEntity, OPackage, OPackageInstantiation } from './parser/objects';
import { SettingsGetter, VhdlLinter } from './vhdl-linter';

export class ProjectParser {

  public cachedFiles: FileCache[] = [];
  public packages: OPackage[] = [];
  public packageInstantiations: OPackageInstantiation[] = [];
  public contexts: OContext[] = [];
  public entities: OEntity[] = [];
  events = new EventEmitter();
  private watchers: FSWatcher[] = [];
  // Constructor can not be async. So constructor is private and use factory to create
  public static async create(workspaces: string[], fileIgnoreRegex: string, settingsGetter: SettingsGetter, disableWatching = false) {
    const projectParser = new ProjectParser(workspaces, fileIgnoreRegex, settingsGetter);
    await projectParser.init(disableWatching);
    return projectParser;
  }
  private constructor(public workspaces: string[], public fileIgnoreRegex: string, public settingsGetter: SettingsGetter) { }
  public addFolders(pathes: string[]) {
    for (const path of pathes) {
      const watcher = watch(path.replace(sep, '/') + '/**/*.vhd?(l)', { ignoreInitial: true });
      watcher.on('add', async (path) => {
        const cachedFile = await (FileCache.create(path, this));
        this.cachedFiles.push(cachedFile);
        this.flattenProject();
        this.events.emit('change', 'add', path);
      });
      watcher.on('change', async (path) => {
        // console.log('change', path);
        const cachedFile = process.platform === 'win32'
          ? this.cachedFiles.find(cachedFile => cachedFile.path.toLowerCase() === path.toLowerCase())
          : this.cachedFiles.find(cachedFile => cachedFile.path === path);
        if (cachedFile) {
          await cachedFile.parse();
          this.flattenProject();
          this.events.emit('change', 'change', path);
        } else {
          console.error('modified file not found', path);
        }
      });
      watcher.on('unlink', path => {
        const cachedFileIndex = this.cachedFiles.findIndex(cachedFile => cachedFile.path === path);
        if (cachedFileIndex > -1) {
          this.cachedFiles.splice(cachedFileIndex, 1);
          this.flattenProject();
          this.events.emit('change', 'unlink', path);
        }
      });
      this.watchers.push(watcher);
    }
  }
  private async init(disableWatching: boolean) {
    const files = new Set<string>();
    await Promise.all(this.workspaces.map(async (directory) => {
      const directories = await this.parseDirectory(directory);
      return (await Promise.all(directories.map(file => promises.realpath(file)))).forEach(file => files.add(file));
    }));
    (await this.parseDirectory(join(cwd(), 'ieee2008'))).forEach(file => files.add(file));

    for (const file of files) {
      const cachedFile = await FileCache.create(file, this);
      this.cachedFiles.push(cachedFile);
    }
    this.cachedFiles.sort((a, b) => b.lintingTime - a.lintingTime);
    // console.log('Times: \n' + this.cachedFiles.slice(0, 10).map(file => `${file.path}: ${file.lintingTime}ms`).join('\n'));
    this.flattenProject();
    if (!disableWatching) {
      this.addFolders(this.workspaces);
    }
  }
  async stop() {
    for (const watcher of this.watchers) {
      await watcher.close();
    }
  }
  private async parseDirectory(directory: string): Promise<string[]> {
    const files: string[] = [];
    const entries = await promises.readdir(directory);
    const ignoreRegex = this.fileIgnoreRegex.trim().length > 0 ? new RegExp(this.fileIgnoreRegex) : null;

    // const entries = await promisify(directory.getEntries)()
    await Promise.all(entries.map(async entry => {
      try {
        const filePath = directory + sep + entry;
        const fileStat = await promises.stat(filePath);
        if (fileStat.isFile()) {
          if (entry.match(/\.vhdl?$/i) && (ignoreRegex === null || !filePath.match(ignoreRegex))) {
            files.push(filePath);
          }
        } else {
          files.push(... await this.parseDirectory(filePath));
        }
      } catch (e) {
        console.error(e);
      }
    }));
    return files;
  }
  private flattenProject() {
    this.packages = [];
    this.packageInstantiations = [];
    this.entities = [];
    for (const cachedFile of this.cachedFiles) {
      this.entities.push(...cachedFile.entities);
      if (cachedFile.packages) {
        this.packages.push(...cachedFile.packages);
      }
      if (cachedFile.packageInstantiations) {
        this.packageInstantiations.push(...cachedFile.packageInstantiations);
      }
      if (cachedFile.contexts) {
        this.contexts.push(...cachedFile.contexts);
      }
    }
  }

}

export class FileCache {
  packages?: OPackage[];
  packageInstantiations?: OPackageInstantiation[];
  contexts: OContext[] = [];
  entities: OEntity[] = [];
  text: string;
  linter: VhdlLinter;
  lintingTime: number;
  // Constructor can not be async. So constructor is private and use factory to create
  public static async create(path: string, projectParser: ProjectParser) {
    const cache = new FileCache(path, projectParser);
    await cache.parse();
    return cache;
  }
  private constructor(public path: string, public projectParser: ProjectParser) {
  }
  async parse() {
    this.text = await promises.readFile(this.path, { encoding: 'utf8' });
    this.linter = new VhdlLinter(this.path, this.text, this.projectParser, this.projectParser.settingsGetter);
    this.packages = this.linter.file.packages.filter((p): p is OPackage => p instanceof OPackage);
    this.packageInstantiations = this.linter.file.packageInstantiations;
    this.entities = this.linter.file.entities;
    this.contexts = this.linter.file.contexts;
  }

}