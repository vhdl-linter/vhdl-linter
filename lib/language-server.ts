import {
  CodeAction, createConnection, DefinitionLink, DidChangeConfigurationNotification, InitializeParams, LSPErrorCodes, Position, ProposedFeatures, ResponseError, TextDocuments, TextDocumentSyncKind
} from 'vscode-languageserver/node';
import { TextDocument } from 'vscode-languageserver-textdocument';
import { URI } from 'vscode-uri';
import { getCompletions } from './languageFeatures/completion';
import { handleDocumentFormatting } from './languageFeatures/documentFormatting';
import { documentHighlightHandler } from './languageFeatures/documentHighlightHandler';
import { findReferencesHandler, prepareRenameHandler, renameHandler } from './languageFeatures/findReferencesHandler';
import { foldingHandler } from './languageFeatures/folding';
import { handleReferences } from './languageFeatures/references';
import { handleOnWorkspaceSymbol } from './languageFeatures/workspaceSymbols';
import { OComponent, OFile, OInstantiation, OUseClause } from './parser/objects';
import { ProjectParser } from './project-parser';
import { VhdlLinter } from './vhdl-linter';
import { handleSemanticTokens, semanticTokensLegend } from './languageFeatures/semanticTokens';
import { existsSync } from 'fs';
import { ISettings, defaultSettings } from './settings';
import { CancelationError, CancelationObject } from './server-objects';
import { getDocumentSymbol } from './languageFeatures/documentSymbol';
import { findObjectFromPosition } from './languageFeatures/findObjectFromPosition';
import { implementsIHasDefinitions } from './parser/interfaces';

// Create a connection for the server. The connection auto detected protocol
// Also include all preview / proposed LSP features.
export const connection = createConnection(ProposedFeatures.all);


// Create a simple text document manager. The text document manager
// supports full document sync only
export const documents: TextDocuments<TextDocument> = new TextDocuments(TextDocument);

let hasWorkspaceFolderCapability = false;
// let hasDiagnosticRelatedInformationCapability: boolean = false;
let projectParser: ProjectParser;
let rootUri: string | undefined;




let globalSettings: ISettings = defaultSettings;
let hasConfigurationCapability = false;
// Cache the settings of all open documents
const documentSettings: Map<string, ISettings> = new Map();
connection.onDidChangeConfiguration(change => {

  if (hasConfigurationCapability) {
    // Reset all cached document settings
    documentSettings.clear();
  } else {
    globalSettings = <ISettings>(
      (change.settings.VhdlLinter || defaultSettings)
    );
  }

  // Revalidate all open text documents
  documents.all().forEach((textDocument) => validateTextDocument(textDocument));
});
export async function getDocumentSettings(resource: string): Promise<ISettings> {
  if (!hasConfigurationCapability) {
    return Promise.resolve(globalSettings);
  }
  let result = documentSettings.get(resource);
  if (!result) {
    result = await connection.workspace.getConfiguration({
      scopeUri: resource,
      section: 'VhdlLinter'
    });
  }
  if (!result) {
    return defaultSettings;
  }
  documentSettings.set(resource, result);
  return result;
}
connection.onInitialize((params: InitializeParams) => {
  const capabilities = params.capabilities;
  hasWorkspaceFolderCapability =
    capabilities.workspace?.workspaceFolders ?? false;
  if (params.rootUri) {
    rootUri = params.rootUri;
  }
  hasConfigurationCapability = capabilities.workspace?.configuration ?? false;
  return {
    capabilities: {
      textDocumentSync: TextDocumentSyncKind.Full,
      codeActionProvider: true,
      // Tell the client that the server supports code completion
      completionProvider: {
        resolveProvider: false
      },
      semanticTokensProvider: {
        legend: semanticTokensLegend,
        range: false,
        full: {
          delta: false
        }
      },

      documentSymbolProvider: true,
      definitionProvider: true,
      hoverProvider: true,
      documentFormattingProvider: {
        workDoneProgress: true
      },
      referencesProvider: true,
      foldingRangeProvider: true,
      documentHighlightProvider: true,
      executeCommandProvider: { commands: ['vhdl-linter:lsp-command'] },
      codeLensProvider: {
        resolveProvider: true
      },
      renameProvider: {
        prepareProvider: true
      },
      workspaceSymbolProvider: true,
    }
  };
});
export const initialization = new Promise<void>(resolve => {
  connection.onInitialized(async () => {
    if (hasConfigurationCapability) {
      // Register for all configuration changes.
      connection.client.register(DidChangeConfigurationNotification.type, undefined);
    }
    const configuration = (await connection.workspace.getConfiguration({
      section: 'VhdlLinter'
    })) as ISettings;

    if (hasWorkspaceFolderCapability) {
      const parseWorkspaces = async () => {
        const workspaceFolders = await connection.workspace.getWorkspaceFolders();
        const folders = (workspaceFolders ?? []).map(workspaceFolder => URI.parse(workspaceFolder.uri).fsPath);
        // console.log(configuration, 'configuration');
        folders.push(...configuration?.paths?.additional?.filter(existsSync) ?? []);
        projectParser = await ProjectParser.create(folders, configuration?.paths?.ignoreRegex ?? '', getDocumentSettings);
      };
      await parseWorkspaces();
      connection.workspace.onDidChangeWorkspaceFolders(async event => {
        projectParser.addFolders(event.added.map(folder => URI.parse(folder.uri).fsPath));
        connection.console.log('Workspace folder change event received.');
      });
    } else {
      const folders = [];
      if (rootUri) {
        folders.push(URI.parse(rootUri).fsPath);
      }
      // console.log('folders', folders);
      projectParser = await ProjectParser.create(folders, configuration.paths.ignoreRegex, getDocumentSettings);
    }
    for (const textDocument of documents.all()) {
      await validateTextDocument(textDocument);
    }
    projectParser.events.on('change', () => {
      // console.log('projectParser.events.change', new Date().getTime(), ... args);
      documents.all().forEach((textDocument) => validateTextDocument(textDocument));
    });
    const timeoutMap = new Map<string, NodeJS.Timeout>();
    const cancelationMap = new Map<string, CancelationObject>();
    documents.onDidChangeContent(change => {
      const oldTimeout = timeoutMap.get(change.document.uri);
      if (oldTimeout !== undefined) {
        clearTimeout(oldTimeout);
      }
      function handleChange() {
        const cancelationObject = cancelationMap.get(change.document.uri);
        if (cancelationObject) {
          cancelationObject.canceled = true;
        }
        const newCancelationObject = {
          canceled: false
        };
        validateTextDocument(change.document, newCancelationObject);
        cancelationMap.set(change.document.uri, newCancelationObject);
      }
      if (change.document.version === 1) { // Document was initially opened. Do not delay.
        handleChange();
      } else {
        timeoutMap.set(change.document.uri, setTimeout(handleChange, 100));
      }
      // const lintingTime = Date.now() - date;
      // console.log(`${change.document.uri}: ${lintingTime}ms`);

    });
    resolve();
  });
});
documents.onDidClose(change => {
  connection.sendDiagnostics({ uri: change.document.uri, diagnostics: [] });
});
// The content of a text document has changed. This event is emitted
// when the text document first opened or when its content has changed.
export const linters = new Map<string, VhdlLinter>();
export const lintersValid = new Map<string, boolean>();
async function validateTextDocument(textDocument: TextDocument, cancelationObject: CancelationObject = { canceled: false }): Promise<void> {
  // console.log(textDocument.uri);
  // console.profile('a');
  // let start = Date.now();
  const vhdlLinter = new VhdlLinter(URI.parse(textDocument.uri).fsPath, textDocument.getText(), projectParser, getDocumentSettings, false, cancelationObject);
  if (vhdlLinter.parsedSuccessfully || typeof linters.get(textDocument.uri) === 'undefined') {
    linters.set(textDocument.uri, vhdlLinter);
    lintersValid.set(textDocument.uri, true);
  } else {
    lintersValid.set(textDocument.uri, false);
  }
  // console.log(`parsed for: ${Date.now() - start} ms.`);
  // start = Date.now();
  try {
    const diagnostics = await vhdlLinter.checkAll();
    diagnostics.forEach((diag) => diag.source = 'vhdl-linter');
    // console.log(`checked for: ${Date.now() - start} ms.`);
    // start = Date.now();
    // console.profileEnd('a');
    connection.sendDiagnostics({ uri: textDocument.uri, diagnostics });
    // console.log(`send for: ${Date.now() - start} ms.`);

  } catch (err) {
    // Ignore cancelled
    if (!(err instanceof CancelationError)) {
      throw err;
    }
  }

}

connection.onDidChangeWatchedFiles(() => {
  // Monitored files have change in VS Code
  connection.console.log('We received an file change event');
});
connection.onCodeAction(async (params): Promise<CodeAction[]> => {
  await initialization;
  const linter = linters.get(params.textDocument.uri);
  if (!linter) {
    return [];
  }
  // linter.codeActionEvent.emit()
  const actions = [];
  for (const diagnostic of params.context.diagnostics) {
    if (typeof diagnostic.code === 'number') {
      const callback = linter.diagnosticCodeActionRegistry[diagnostic.code];
      if (typeof callback === 'function') {
        actions.push(...callback(params.textDocument.uri));
      }
    } else if (typeof diagnostic.code === 'string') {
      const codes = diagnostic.code.split(';').map(a => parseInt(a));
      for (const code of codes) {
        const callback = linter.diagnosticCodeActionRegistry[code];
        if (typeof callback === 'function') {
          actions.push(...callback(params.textDocument.uri));
        }

      }
    }
  }
  return actions;
});
connection.onDocumentSymbol(async (params) => {
  await initialization;
  const linter = linters.get(params.textDocument.uri);
  if (!linter) {
    return [];
  }
  return getDocumentSymbol(linter);
});
interface IFindDefinitionParams {
  textDocument: {
    uri: string
  };
  position: Position;
}
const findDefinitions = async (params: IFindDefinitionParams): Promise<(DefinitionLink & { text: string })[]> => {
  await initialization;
  const linter = linters.get(params.textDocument.uri);
  if (!linter) {
    return [];
  }

  const candidates = findObjectFromPosition(linter, params.position);
  return candidates.flatMap(candidate => {
    if (implementsIHasDefinitions(candidate) && candidate.definitions) {
      return candidate.definitions.map(definition => {
        return {
          targetRange: definition.range.copyExtendBeginningOfLine().getLimitedRange(10),
          targetSelectionRange: definition.lexerToken?.range ?? definition.range.copyExtendBeginningOfLine().getLimitedRange(1),
          text: definition.rootFile.originalText,
          targetUri: URI.file(definition.rootFile.file).toString()
        };
      });
    } else {
      return [];
    }
  });
};
const findBestDefinition = async (params: IFindDefinitionParams) => {
  const definitions = await findDefinitions(params);
  if (definitions.length === 0) {
    return null;
  }
  return definitions[0];
};

connection.onHover(async (params, token) => {
  await initialization;
  if (token.isCancellationRequested) {
    console.log('hover canceled');
    return new ResponseError(LSPErrorCodes.RequestCancelled, 'canceled');
  }
  const definition = await findBestDefinition(params);
  if (definition === null) {
    return null;
  }
  const lines = definition.text.split('\n').slice(definition.targetRange.start.line, definition.targetRange.end.line + 1);
  if (definition.targetRange.start.line === definition.targetRange.end.line) {
    lines[0] = lines[0].substring(definition.targetRange.start.character, definition.targetRange.end.character);
  } else {
    lines[0] = lines[0].substring(definition.targetRange.start.character);
    lines[lines.length - 1] = lines[lines.length - 1].substring(0, definition.targetRange.end.character);
  }
  return {
    contents: {
      language: 'vhdl',
      value: lines.join('\n')
    }
  };
});
connection.onDefinition(findDefinitions);
connection.onCompletion(async (params, cancelationToken) => {
  await initialization;
  const linter = linters.get(params.textDocument.uri);
  if (!linter) {
    return [];
  }
  if (cancelationToken.isCancellationRequested) {
    return [];
  }
  return getCompletions(linter, params);
});
connection.onReferences(handleReferences);
connection.onPrepareRename(prepareRenameHandler);
connection.onRenameRequest(renameHandler);
connection.onDocumentFormatting(handleDocumentFormatting);
connection.onFoldingRanges(foldingHandler);
connection.onDocumentHighlight(async (params, cancelationToken) => {
  await initialization;
  const linter = linters.get(params.textDocument.uri);
  if (!linter) {
    return [];
  }
  if (cancelationToken.isCancellationRequested) {
    return [];
  }
  return documentHighlightHandler(linter, params);

});
connection.onReferences(findReferencesHandler);
connection.onWorkspaceSymbol(params => handleOnWorkspaceSymbol(params, projectParser));
// connection.on
// eslint-disable-next-line @typescript-eslint/no-explicit-any
connection.onRequest('vhdl-linter/listing', async (params: any) => {
  await initialization;
  const textDocumentUri = params.textDocument.uri;
  const linter = linters.get(textDocumentUri);
  if (typeof linter === 'undefined') {
    return;
  }
  const files: OFile[] = [];
  const unresolved: string[] = [];

  function addUnresolved(name: string) {
    if (unresolved.findIndex(search => search === name) === -1) {
      unresolved.push(name);
    }
  }

  async function parseTree(file: OFile) {
    const index = files.findIndex(fileSearch => fileSearch?.file === file?.file);
    if (index === -1) {
      files.push(file);
    } else {
      // push the file to the back to have correct compile order
      files.push(files.splice(index, 1)[0]);
    }

    for (const obj of file.objectList) {
      let found: OFile | undefined = undefined;
      if (obj instanceof OInstantiation) {
        if (obj.type === 'entity') {
          if (obj.definitions?.length > 0 && obj.definitions[0].parent instanceof OFile && obj.definitions[0].parent.entities[0] !== undefined) { // TODO: Fix me better
            found = obj.definitions[0].parent;
          } else {
            addUnresolved(`${obj.library}.${obj.componentName.text}`);
          }
        } else if (obj.type === 'component') {
          if (obj.definitions?.length > 0
            && obj.definitions[0] instanceof OComponent && obj.definitions[0].definitions?.length > 0
            && obj.definitions[0].definitions[0].parent instanceof OFile) {
            found = obj.definitions[0].definitions[0].parent;
          } else {
            addUnresolved(obj.componentName.text);
          }
        }
      } else if (obj instanceof OUseClause) {
        // do not generate file listings for ieee files
        if (obj.library?.getLText() === 'ieee' || obj.library?.getLText() === 'std') {
          continue;
        }
        const matchingPackages = projectParser.packages.filter(pkg => pkg.lexerToken.getLText() === obj.packageName.getLText());
        if (matchingPackages.length > 0) {
          found = matchingPackages[0].parent;
        }
      }

      if (found) {
        const vhdlLinter = new VhdlLinter(found.file, found.originalText, projectParser, getDocumentSettings);
        await vhdlLinter.checkAll();
        await parseTree(vhdlLinter.file);
      }
    }

  }

  await parseTree(linter.file);
  const filesList = files.reverse().map(file => file.file.replace((rootUri ?? '').replace('file://', ''), '')).join(`\n`);
  const unresolvedList = unresolved.join('\n');
  return `files:\n${filesList}\n\nUnresolved instantiations:\n${unresolvedList}`;
});
connection.languages.semanticTokens.on(handleSemanticTokens);
documents.listen(connection);

// Listen on the connection
connection.listen();
