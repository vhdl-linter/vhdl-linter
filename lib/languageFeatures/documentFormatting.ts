import { exec, spawn } from 'child_process';
import { promises } from 'fs';
import { tmpdir } from 'os';
import { sep } from 'path';
import { platform } from 'process';
import { fileURLToPath } from 'url';
import { promisify } from 'util';
import { CancellationToken, DocumentFormattingParams, LSPErrorCodes, Range, ResponseError, TextEdit, WorkDoneProgressReporter } from 'vscode-languageserver';
import { attachWorkDone } from 'vscode-languageserver/lib/common/progress';
import { connection, documents } from '../language-server';
import { getRootDirectory, joinURL } from '../project-parser';
// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-argument
const nullProgressReporter = attachWorkDone(undefined as any, /* params */ undefined);
async function _getProgressReporter(reporter: WorkDoneProgressReporter, title: string) {
  // This is a bit ugly, but we need to determine whether the provided reporter
  // is an actual client-side progress reporter or a dummy (null) progress reporter
  // created by the LSP library. If it's the latter, we'll create a server-initiated
  // progress reporter.
  if (reporter.constructor !== nullProgressReporter.constructor) {
    return reporter;
  }

  const serverInitiatedReporter = await connection.window.createWorkDoneProgress();
  serverInitiatedReporter.begin(
    title
  );

  return serverInitiatedReporter;
}
const execP = promisify(exec);
export async function handleDocumentFormatting(params: DocumentFormattingParams, token: CancellationToken, workDoneProgress: WorkDoneProgressReporter): Promise<TextEdit[] | null | ResponseError> {
  const document = documents.get(params.textDocument.uri);
  if (typeof document === 'undefined') {
    return null;
  }
  const text = document.getText();
  const path = await promises.mkdtemp(tmpdir() + sep);
  const tmpFile = path + sep + 'beautify';
  await promises.writeFile(tmpFile, text);
  const rootDir = getRootDirectory();
  const emacsScripts = joinURL(rootDir, 'emacs', 'emacs-vhdl-formatting-script.lisp');
  const emacsLoadPath = joinURL(rootDir, 'emacs');
  const numSpaces = typeof params.options.tabSize === 'number' ? params.options.tabSize : 2;
  let foundEmacs = false;
  if (platform === 'win32') {
    try {
      await execP(`emacs --batch`);
      foundEmacs = true;
    } catch (e) {
    }
  } else {
    try {
      await execP(`command -v emacs`);
      foundEmacs = true;
    } catch (e) {
    }
  }
  if (!foundEmacs) {
    connection.window.showErrorMessage('vhdl-linter is using emacs for formatting. Install emacs for formatting to work.');
    return null;

  }
  const controller = new AbortController();
  const { signal } = controller;
  token.onCancellationRequested(() => {
    controller.abort();
  });
  const progress = await _getProgressReporter(workDoneProgress, 'Emacs Formatter running');

  if (platform === 'win32') {
    progress.report(0, "Emacs Formatter running");

    const cmd = `emacs --batch --eval "(setq-default vhdl-basic-offset ${numSpaces})" ` +
      // The vhdl-mode from the folder does not work with win for some reason
      // `--eval "(setq load-path (cons (expand-file-name \\"${fileURLToPath(emacsLoadPath)}\\") load-path))" ` +
      ` -l ${fileURLToPath(emacsScripts)} -f vhdl-batch-indent-region ${tmpFile}`
    try {
      await execP(cmd);
    } catch (e) {
      progress.done();

      if (e.name === "AbortError") {
        return new ResponseError(LSPErrorCodes.RequestCancelled, 'canceled');
      }
      throw e;
    }
    progress.done();

  } else {
    try {
      await new Promise<void>((resolve, reject) => {
        const args = ['-c', `emacs --batch --eval "(setq-default vhdl-basic-offset ${numSpaces})" ` +
          `--eval "(setq load-path (cons (expand-file-name \\"${fileURLToPath(emacsLoadPath)}\\") load-path))" ` +
          ` -l ${fileURLToPath(emacsScripts)} -f vhdl-batch-indent-region ${tmpFile}`];
        const emacs = spawn('sh', args, { signal });

        emacs.stderr.on('data', (data: Buffer) => {
          const match = data.toString().match(/(\d+)%/) as [string, string] | null;
          if (match) {
            progress.report(parseInt(match[1]), data.toString());
          }
        });
        emacs.on('error', err => reject(err));
        emacs.on('close', code => {
          if (code === 0) {
            resolve();
          }
          reject(code);
        });
      });

      progress.done();
    } catch (e) {
      progress.done();

      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
      if (e.name === "AbortError") {
        return new ResponseError(LSPErrorCodes.RequestCancelled, 'canceled');
      }
      throw e;
    }
  }
  // Emacs VHDL mode hard tabs seem to be broken.
  // Replace spaces in beginning with tabs
  let newText = await promises.readFile(tmpFile, { encoding: 'utf8' });
  if (params.options.insertSpaces === false) {
    const re = /^( +)/gm;
    console.log(re);
    newText = newText.replaceAll(re,
      match => '\t'.repeat(Math.ceil(match.length / numSpaces))
    );
  }
  return [{
    range: Range.create(document.positionAt(0), document.positionAt(text.length)),
    newText
  }];
}