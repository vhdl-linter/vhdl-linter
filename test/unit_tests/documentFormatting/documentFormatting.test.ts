/* eslint-disable @typescript-eslint/no-unnecessary-type-assertion */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { beforeEach, expect, jest, test } from '@jest/globals';
import * as child_process from 'child_process';
import * as fs from 'fs';
import * as process from 'process';
import { CancellationTokenSource, Position } from 'vscode-languageserver';
import { attachWorkDone } from 'vscode-languageserver/lib/common/progress';
import { connection } from '../../../lib/language-server';
import { handleDocumentFormatting } from '../../../lib/languageFeatures/documentFormatting';
import * as PP from '../../../lib/project-parser';

const text = 'dummy Text';
const expectedTest = 'expected dummy result';
const nullProgressReporter = attachWorkDone(undefined as any, /* params */ undefined);
beforeEach(() => {
  jest.clearAllMocks();
});
// jest.mock('vscode-languageserver');
jest.mock('../../../lib/language-server', () => {
  return {
    documents: {
      get: () => {
        return {
          getText: jest.fn(() => text),
          positionAt: jest.fn(() => Position.create(0, 0))
        };
      },
    },
    connection: {
      window: {
        createWorkDoneProgress: jest.fn(() => ({
          begin: jest.fn(),
          done: jest.fn(),
          report: jest.fn()
        }))
      }
    }
  };
});
const mockConnection = jest.mocked(connection);
jest.mock('child_process');
jest.mock('fs', () => {
  const originalModule = jest.requireActual('fs') as any;
  return {
    __esModule: true,
    ...originalModule,
    promises: {
      readFile: jest.fn(),
      mkdtemp: jest.fn(),
      writeFile: jest.fn()
    }
  };
});
jest.mock('process', () => {
  return {
    platform: ''
  };
});
jest.mock('../../../lib/project-parser', () => {
  const originalModule = jest.requireActual('../../../lib/project-parser') as any;
  return {
    __esModule: true,
    ...originalModule,
    getRootDirectory: jest.fn()
  };
});

const mockFs = jest.mocked(fs);
const mockChild_process = jest.mocked(child_process);
const mockProcess = jest.mocked(process);
const mockProjectParser = jest.mocked(PP);
test.each([
  'win32',
  'linux',
])('Testing formatting call on platform %s', async (platform) => {

  (mockProcess.platform as string) = platform;

  const uri = 'file:///dummy.vhd';
  const cancellationTokenSource = new CancellationTokenSource();
  mockProjectParser.getRootDirectory.mockImplementation(() => new URL('file:///dummyRoot'));
  mockChild_process.exec.mockImplementationOnce(((_command: string, callback: () => void) => {
    callback();
  }) as any);
  if (platform === 'win32') {
    mockChild_process.exec.mockImplementationOnce(((_command: string, callback: () => void) => {
      callback();
    }) as any);
  } else {
    mockChild_process.spawn.mockImplementationOnce(() => {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-return
      return {
        on: (eventName: string, callback: (par: any) => void) => {
          if (eventName === 'close') {
            callback(0);
          }
        },
        stderr: {
          on: jest.fn()
        }
      } as any;
    });
  }

  mockFs.promises.readFile.mockResolvedValueOnce(expectedTest);
  mockFs.promises.mkdtemp.mockResolvedValueOnce('tmpName');
  mockFs.promises.writeFile.mockResolvedValueOnce();
  const formatting = await handleDocumentFormatting({
    textDocument: {
      uri: uri
    },
    options: {
      tabSize: 2,
      insertSpaces: true
    }
  }, cancellationTokenSource.token, nullProgressReporter);
  if (platform === 'win32') {
    expect(mockChild_process.exec.mock.calls[0]?.[0]).toMatchInlineSnapshot(`"emacs --batch"`);
    expect(mockChild_process.exec.mock.calls[1]?.[0]).toMatchInlineSnapshot(`"emacs --batch --eval "(setq-default vhdl-basic-offset 2)"  -l /dummyRoot/emacs/emacs-vhdl-formatting-script.lisp -f vhdl-batch-indent-region tmpName/beautify"`);
  } else {
    expect(mockChild_process.exec.mock.calls[0]?.[0]).toMatchInlineSnapshot(`"command -v emacs"`);
    expect(mockChild_process.spawn.mock.calls[0]?.[0]).toMatchInlineSnapshot(`"sh"`);
    expect(mockChild_process.spawn.mock.calls[0]?.[1]).toMatchInlineSnapshot(`
[
  "-c",
  "emacs --batch --eval "(setq-default vhdl-basic-offset 2)" --eval "(setq load-path (cons (expand-file-name \\"/dummyRoot/emacs\\") load-path))"  -l /dummyRoot/emacs/emacs-vhdl-formatting-script.lisp -f vhdl-batch-indent-region tmpName/beautify",
]
`);

  }

  expect(formatting).not.toBeNull();
  expect(formatting![0]!.newText).toBe(expectedTest);
  expect(formatting).toMatchSnapshot();
  expect((mockConnection.window.createWorkDoneProgress.mock.results[0]!.value as any).done).toHaveBeenCalled();
});