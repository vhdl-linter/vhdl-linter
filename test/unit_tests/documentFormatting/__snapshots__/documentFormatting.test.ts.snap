// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing formatting call on win32 3`] = `
[
  {
    "newText": "expected dummy result",
    "range": {
      "end": {
        "character": 0,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
  },
]
`;
exports[`Testing formatting call on linux (non win32) 3`] = `
[
  {
    "newText": "expected dummy result",
    "range": {
      "end": {
        "character": 0,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
  },
]
`;
